core/src/main/java/org/kiwix/kiwixmobile/core/history/viewmodel/HistoryViewModel.kt

```diff
@@ -0,0 +1,164 @@
+  private var isInSelectionMode = false
+
+  init {
+    compositeDisposable.addAll(viewStateReducer(), actionMapper(), updateSelectionMode())
+  }
+
+  private fun updateSelectionMode() =
+    unselectAllItems.subscribe({ isInSelectionMode = it }, Throwable::printStackTrace)
+
+  private fun viewStateReducer() = Flowable.combineLatest(
+    currentBook,
+    searchResults(),
+    unselectAllItems,
+    filter,
+    showAllSwitchToggle,
+    Function5(::updateResultsState)
+  )
+    .subscribe(state::postValue, Throwable::printStackTrace)
+
+  private fun searchResults(): Flowable<List<HistoryListItem>> =
+    Flowable.combineLatest(
+      filter,
+      showAllSwitchToggle,
+      historyDao.history(),
+      Function3(::searchResults)
+    )
+
+  private fun searchResults(
+    searchString: String,
+    showAllToggle: Boolean,
+    historyList: List<HistoryListItem>
+  ): List<HistoryListItem> =
+    historyList.filterIsInstance<HistoryItem>()
+      .filter { h ->
+        h.historyTitle.contains(searchString, true) &&
+          (h.zimName == zimReaderContainer.name || showAllToggle)
+      }
+
+  private fun updateResultsState(
+    currentBook: String,
+    historyBookResults: List<HistoryListItem>,
+    unselectAllItems: Boolean,
+    searchString: String,
+    showAllSwitchOn: Boolean
+  ): State {
+    if(unselectAllItems){
+      historyBookResults.filterIsInstance<HistoryItem>().forEach { it.isSelected = false }
+    }
+    return Results(searchString, historyBookResults, showAllSwitchOn, currentBook)
+  }
+
+  override fun onCleared() {
+    compositeDisposable.clear()
+    super.onCleared()
+  }
+
+  private fun actionMapper() = actions.map {
+    when (it) {
+      ExitHistory -> effects.offer(Finish)
+      is Filter -> filter.offer(it.searchTerm)
+      is ToggleShowHistoryFromAllBooks -> showAllSwitchToggle.offer(it.isChecked)
+      is ToggleShowAllHistoryAvailability ->
+        effects.offer(ToggleShowAllHistorySwitchAvailability(it.showAllHistorySwitch))
+      is CreatedWithIntent -> filter.offer(it.searchTerm)
+      is ConfirmedDelete -> deleteItemAndShowToast(it)
+      is OnItemLongClick -> selectItemAndOpenSelectionMode(it.historyItem)
+      is OnItemClick -> appendItemToSelectionOrOpenIt(it)
+      is DeleteHistoryItems -> historyDao.deleteHistory(it.itemsToDelete)
+      ReceivedPromptForSpeechInput -> effects.offer(StartSpeechInput(actions))
+      ExitActionModeMenu -> unselectAllItems.offer(true)
+      StartSpeechInputFailed -> effects.offer(ShowToast(R.string.speech_not_supported))
+    }
+  }.subscribe({}, Throwable::printStackTrace)
+
+
+  private fun selectItemAndOpenSelectionMode(historyItem: HistoryItem) {
+    historyItem.isSelected = true
+    unselectAllItems.offer(false)
+  }
+
+  private fun isInSelctionMode(): Boolean {
+    return state
+      .value
+      ?.historyItems
+      ?.filterIsInstance<HistoryItem>()
+      ?.filter { it.isSelected }
```

Comment:
`?.any {condition} ==true` is better and short circuiting

Reply:
Nice! I think AS can recognize that type of simplification in some cases but apparently not this one.